internal/controller/observedobjectcollection/reconciler.go

```diff
@@ -0,0 +1,318 @@
+	}
+
+	if err := clusterClient.List(ctx, objects, &client.ListOptions{LabelSelector: selector, Namespace: c.Spec.ObservedObjects.Namespace}); err != nil {
+		werr := errors.Wrap(err, "error listing objects")
+		c.Status.SetConditions(xpv1.ReconcileError(werr))
+		_ = r.client.Status().Update(ctx, c)
+		return ctrl.Result{}, werr
+	}
+
+	// Create observed-only Objects for all found items.
+	refs := sets.New[v1alpha1.ObservedObjectReference]()
+	for i := range objects.Items {
+		o := objects.Items[i]
+		log.Debug("creating observed object for the matched item", "gvk", o.GroupVersionKind(), "name", o.GetName())
+		name, err := r.observedObjectName(c, &o)
+		if err != nil {
+			werr := errors.Wrapf(err, "error generating name for observed object, matched object: %v", o)
+			c.Status.SetConditions(xpv1.ReconcileError(werr))
+			_ = r.client.Status().Update(ctx, c)
+			return ctrl.Result{}, werr
+		}
+
+		// Create patch
+		po, err := observedObjectPatch(name, o, c)
+		if err != nil {
+			werr := errors.Wrapf(err, "error generating patch for matched object %v", o)
+			c.Status.SetConditions(xpv1.ReconcileError(werr))
+			_ = r.client.Status().Update(ctx, c)
+			return ctrl.Result{}, werr
+		}
+		if l := po.GetLabels(); l != nil {
+			l[membershipLabelKey] = mlv
+			po.SetLabels(l)
+		} else {
+			po.SetLabels(map[string]string{membershipLabelKey: mlv})
+		}
+		if err := r.client.Patch(ctx, po, client.Apply, fieldOwner, client.ForceOwnership); err != nil {
+			werr := errors.Wrap(err, "cannot create observed object")
+			c.Status.SetConditions(xpv1.ReconcileError(werr))
+			_ = r.client.Status().Update(ctx, c)
+			return ctrl.Result{}, werr
+		}
+
+		log.Debug("created observed object", "name", po.GetName())
+		refs.Insert(v1alpha1.ObservedObjectReference{Name: name})
+	}
+
+	// Remove collection members that either do not exist anymore or are no match.
+	ol := &v1alpha2.ObjectList{}
+	if err := r.client.List(ctx, ol, client.MatchingLabels(map[string]string{membershipLabelKey: mlv})); err != nil {
+		werr := errors.Wrapf(err, "cannot list members matching label %v=%v", membershipLabelKey, mlv)
+		c.Status.SetConditions(xpv1.ReconcileError(werr))
+		_ = r.client.Status().Update(ctx, c)
+		return ctrl.Result{}, werr
+	}
+
+	oldRefs := sets.New[v1alpha1.ObservedObjectReference](c.Status.ObjectRefs...)
+	for i := range ol.Items {
+		name := ol.Items[i].Name
+		if refs.Has(v1alpha1.ObservedObjectReference{Name: name}) {
+			continue
+		}
+		log.Debug("Removing", "name", name)
+		if err := r.client.Delete(ctx, &ol.Items[i]); err != nil {
+			werr := errors.Wrapf(err, "cannot delete observed object %s", name)
+			c.Status.SetConditions(xpv1.ReconcileError(werr))
+			_ = r.client.Status().Update(ctx, c)
+			return ctrl.Result{}, werr
+		}
+	}
+	c.Status.SetConditions(xpv1.ReconcileSuccess(), xpv1.Available())
+
+	// Update references only if more or less objects got matched since the last reconciliation.
+	if !refs.Equal(oldRefs) {
```

Comment:
nit: 

```suggestion
	if !refs.Equal(sets.New[v1alpha1.ObservedObjectReference](c.Status.ObjectRefs...)) {
```

and the above line, `oldRefs := sets.New[v1alpha1.ObservedObjectReference](c.Status.ObjectRefs...)` could be removed. This is the only place it is used.

Reply:
Given that actually `status.objectRefs` is not needed because there is a common object label, and in the light of current naming debate, I am going to the refs from the status.

Reply:
I have removed `status.objectRefs` fields from CRD.